File: test/test_landmarktransform.py
# LICENSE
# 
# _This file is Copyright 2018 by the Image Processing and Analysis Group (BioImage Suite Team). Dept. of Radiology & Biomedical Imaging, Yale School of Medicine._
# 
# BioImage Suite Web is licensed under the Apache License, Version 2.0 (the "License");
# 
# - you may not use this software except in compliance with the License.
# - You may obtain a copy of the License at [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)
# 
# __Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.__
# 
# ENDLICENSE

import os
import sys
import numpy as np
import unittest
import tempfile
import json
import math

np.set_printoptions(precision=3)
np.set_printoptions(suppress=True)

my_path=os.path.dirname(os.path.realpath(__file__));
sys.path.insert(0,os.path.abspath(my_path+'/../'));
sys.path.insert(0,os.path.abspath(my_path+'/../biswebpython/modules'));

from biswebpython.core.bis_objects import *
import biswebpython.core.bis_baseutils as bis_baseutils;

libbis=bis_baseutils.getDynamicLibraryWrapper();


fname=my_path+'/../test/testdata/pointlocator/brain.json';
file1=open(fname); text1=file1.read(); brain=json.loads(text1); file1.close();
arr=np.asarray(brain['points'],dtype=np.float32);
print(arr.shape);
l=arr.shape[0];
rows=int(l/3);
cols=3
points=np.reshape(arr,[ rows,cols ]);

m_points=np.ones([rows,cols+1],dtype=np.float32);
print('Size=',points.shape);
print('Size=',m_points.shape);
m_points[:,0:3]=points[:,0:3]

print('____________________________________________________');
print('Num points= ',rows, ' file=',fname);
print("Points 0 & 2 = ",points[0,:], points[2,:])
print("MPoints 0 & 2 = ",m_points[0,:], m_points[2,:])

print('____________________________________________________');


matrix_list = [
    [ [  0.9400000000000001,  0.342,  0 , -21.511 ],
      [  -0.342,  0.9400000000000001,  0,  37.295 ],
      [  0 , 0,  1,  0 ],
      [  0 , 0,  0,  1 ]
    ],
    [ [  1.034,  0.376,  0,  -33.662 ],
      [  -0.376,  1.034,  0,  30.225 ],
      [  0 , 0,  1.1,  -9 ],
      [  0  ,0,  0,  1 ]
    ],
    [ [  0.9380000000000001,  0.38,  0,  -25.44 ],
      [  -0.34600000000000003,  1.032,  0,  27.6 ],
      [  0,  0,  1,  0 ],
      [  0,  0,  0,  1 ]
    ]
]

matrices= [ np.zeros([4,4],dtype=np.float32),
            np.zeros([4,4],dtype=np.float32),
            np.zeros([4,4],dtype=np.float32) ];

for i in range(0,3):
    for row in range(0,4):
        for col in range(0,4):
            matrices[i][row][col]=matrix_list[i][row][col];
    print('Matrix ',i+1,'=',matrices[i]);
        
print('____________________________________________________');

t_points=np.transpose(m_points);


class TestPointLocator(unittest.TestCase):

    def test_fit(self):

        passed=0;
        tested=0;
        numtests=3

        for i in range(0,numtests):

            print('Matrices=',matrices[i].shape,t_points.shape);
            warped=np.transpose(np.matmul(matrices[i],t_points))[:,0:3]

            print('In points=', points[0][:], points[4][:]);
            print('In points=', warped[0][:], warped[4][:]);

            out=libbis.computeLandmarkTransformWASM(points,warped,
                                                    { 'mode' : i },0);

            print('Input=\n',matrices[i]);
            print('Output=\n',out);
            print('Difference=\n',abs(matrices[i]-out))
            
            dl=out.flatten()-matrices[i].flatten();
            diff=max(np.amax(dl),-np.amin(dl));
            print('Difference',diff);
            tested+=1
            if (diff<0.1):
                print('_____ P A S S E D ____\n');
                passed=passed+1;
            else:
                print('_____ F A I L E D ____\n');
                
            
            print('____________________________________________________');


        self.assertEqual(passed,tested);
        

    def test_rpmcorrespodence(self):

        print('Matrices=',matrices[2].shape,t_points.shape);
        warped=np.transpose(np.matmul(matrices[2],t_points))[:,0:3]

        small_pts=points[0:20,:]
        small_wrp=warped[0:23,: ]
        
        print('In points=', small_pts[0][:], small_pts[4][:]);
        print('In points=', small_wrp[0][:], small_wrp[4][:]);

        out=libbis.test_rpmCorrespondenceEstimatorWASM(small_pts,small_wrp,
                                                       { 'mode' : 0 },1);
        print(out[7:10,:]);

        out2=libbis.test_rpmCorrespondenceEstimatorWASM(small_pts,small_wrp,
                                                       { 'mode' : 1,
                                                         'temperature' : 6.0,
                                                       },0);
        print(out2[7:10,:]);

        out3=libbis.test_rpmCorrespondenceEstimatorWASM(small_pts,small_wrp,
                                                       { 'mode' : 2,
                                                         'temperature' : 6.0,
                                                       },0);
        print(out3[7:10,:]);
        
        self.assertEqual(True,True);

    def test_rpmsampling(self):

        print('___________ test rpm sampling ');

        points=np.zeros([ 1000,3],dtype=np.float32);
        labels=np.zeros([ 1000],dtype=np.int32);

        for i in range (0,1000):
            if (i<950):
                points[i]=[ i,2*i,0.1*i ]
                labels[i]=0
            else:
                points[i]=[ i,2*i,1000+0.1*i ]
                labels[i]=1
        

        # Hardcode numpoints=54, max=980
        print('---------------------- UnWeighted Sampling ---------------')
        numpoints=54
        step=int(1000/numpoints)
        print('step=',step)
        
        gold=points[0:980:step];
        
        outpts=libbis.test_rpmSamplingWASM(points,labels,{ 'returnlabels' : 0,
                                                           'numpoints' : numpoints,
                                                           'prefsampling' : 1
                                                           },1);

        print('gold=',gold.shape,'\n',gold[0::6]);
        print('outpts=',outpts.shape,'\n',outpts[0::6]);

        dl=gold.flatten()-outpts.flatten();
        diff=max(np.amax(dl),-np.amin(dl));
        print('Difference = ', diff);

        print('---------------------- Weighted Sampling ---------------')
        steps=[17,4]
        p1=points[0:930:17];
        p2=points[950:998:4];
        print('p1=',p1.shape,'p2=',p2.shape);
        goldp=np.concatenate([ p1,p2],axis=0);
        print('goldp=',goldp.shape);
        
        outpts2=libbis.test_rpmSamplingWASM(points,labels,{ 'returnlabels' : 0,
                                                            'numpoints' : 64,
                                                            'prefsampling' : 4, },1);
        
        

        print('goldp=',goldp.shape,'\n',goldp[0::6]);
        print('outpts2=',outpts2.shape,'\n',outpts2[0::6]);

        dl=goldp.flatten()-outpts2.flatten();
        diff2=max(np.amax(dl),-np.amin(dl));
        print('Difference2 = ', diff2);

        self.assertEqual(True,(diff<0.01 and diff2 <0.01));
            
        

        
        

File: test/test_landmarktransform.js
/*  LICENSE
 
 _This file is Copyright 2018 by the Image Processing and Analysis Group (BioImage Suite Team). Dept. of Radiology & Biomedical Imaging, Yale School of Medicine._
 
 BioImage Suite Web is licensed under the Apache License, Version 2.0 (the "License");
 
 - you may not use this software except in compliance with the License.
 - You may obtain a copy of the License at [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)
 
 __Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.__
 
 ENDLICENSE */

/* jshint node:true */
/*global describe, it, before */
"use strict";

require('../config/bisweb_pathconfig.js');
require('bisweb_userpreferences.js').setImageOrientationOnLoad('None');
const assert = require("assert");
const genericio=require('bis_genericio');
const BisWebImage=require('bisweb_image');
const BisWebMatrix=require('bisweb_matrix');
const BisWebLinearTransformation=require('bisweb_lineartransformation');
const path=require('path');
const numeric=require('numeric');
const libbiswasm=require('libbiswasm_wrapper');

numeric.precision=3;


const transform_points=function(source_pts , matrix) {

    console.log('__ Points=',source_pts.getDescription(),'\n\t 4x4 matrix=',matrix.getDescription());
    
    let mat=matrix.getMatrix();
    let arr=source_pts.getDataArray();
    let np=arr.length/3;

    let dim=source_pts.getDimensions();
    let output=new BisWebMatrix();
    output.zero(dim[0],dim[1]);
    let out=output.getDataArray();

    let sumr=[ 0,0,0];
    let sumt=[0,0,0];
    
    for (let i=0;i<np;i++) {
        let pt= [ arr[i*3], arr[i*3+1], arr[i*3+2],1 ];
        for (let row=0;row<=2;row++) {
            for (let col=0;col<=3;col++) {
                out[i*3+row] += mat[row][col] * pt[col];
            }
        }

        if (i == 1200 === 0 ) {
            console.log('Input  Point ',i,'=', pt[0], pt[1],pt[2]);
            console.log('Output Point ',i,'=', out[3*i], out[3*i+1],out[3*i+2]);
        }

        for (let ia=0;ia<=2;ia++) {
            sumr[ia]+=pt[ia];
            sumt[ia]+=out[3*i+ia];
        }
    }

    for (let ia=0;ia<=2;ia++) {
        sumr[ia]=sumr[ia]/np;
        sumt[ia]=sumt[ia]/np;
    }

    console.log('Centroids=',sumr,sumt);
    return output;
    
};


const print_matrices=function(matrices,threshold,names=['','' ]) {

    for (let i=0;i<matrices.length;i++) {
        console.log('_________________________ Matrix= ',names[i]);
        let output='';
        for (let row=0;row<=3;row++) {
            output=output+'[ ';
            for (let col=0;col<=3;col++) {
                
                let v=matrices[i][row][col];
                v=0.001*Math.round(v*1000.0);
                if (Math.abs(v)<threshold)
                    v=0.0;
                output=`${output} ${v} `;
            }
            output=output+']\n';
        }
        console.log(output);
    }

};

describe('Testing landmark transformation creation\n', function() {

    this.timeout(50000);    
    let image = new BisWebImage();
    let imagename=path.resolve(__dirname, 'testdata/avg152T1_LR_nifti.nii.gz');
    let surfacename=path.resolve(__dirname, 'testdata/pointlocator/brain.json');
    let source_pts=new BisWebMatrix();
    
    before( async function() {

        await libbiswasm.initialize();
        await image.load(imagename);
        let obj=await genericio.read(surfacename);
        let points=JSON.parse(obj.data)['points'];
        let nt=points.length;
        let np=Math.round(nt/3);
        console.log('Number of points=',np,np*3,nt, np*3-nt);

        source_pts.zero(np,3);
        let arr=source_pts.getDataArray();
        for (let i=0;i<nt;i++)
            arr[i]=points[i];
        
        return Promise.resolve('done');
    });

    

    it('wasm test Landmark Transform 1',function() {
        
        const pvectors=[ new Float32Array([10.0,0,0,
                                          0,0,20,
                                          1.0,1.0,1.0,
                                          0.0,0.0,0.0]) ,
                        new Float32Array([10.0,0,0,
                                          0,0,20,
                                          1.1,1.1,1.1,
                                          0.0,0.0,0.0]) ,
                        new Float32Array([10.0,0,0,
                                          0,0,20,
                                          1.0,1.1,1.0,
                                          0.0,0.0,0.2]) ];
        
        
        const dim = image.getDimensions(),spa=image.getSpacing();
        console.log('Images = ',dim,spa);

        let maxd=[0,0,0];
        
        for (let mode=0;mode<=2;mode++) {

            let md=mode;
            if (md==2)
                md=3;
            
            console.log('__________________________________________________________');
            console.log('____ Beginning mode=',md);
        
            const tr_js=new BisWebLinearTransformation(3);
            tr_js.setShifts(dim,spa,dim,spa);
            let opts = { rigidOnly : true };
            if (mode >0 ) {
                opts = { rigidOnly : false };
            }
            tr_js.setParameterVector(pvectors[mode],opts);
            console.log('___ pass ',md);
            print_matrices([ tr_js.getMatrix() ],1e-4, [ `original_${md}` ]);
            
            const targ_pts=transform_points(source_pts,tr_js);
            let output_js=libbiswasm.computeLandmarkTransformWASM(source_pts,targ_pts,{ 'mode' : mode },1);


            maxd[mode]=0.0;
            for (let i=0;i<=3;i++) {
                for (let j=0;j<=3;j++) {
                    let v=Math.abs(output_js.getMatrix()[i][j]-tr_js.getMatrix()[i][j]);
                    if (maxd[mode]<v)
                        maxd[mode]=v;
                }
            }
            
            const matrices= [  tr_js.getMatrix(), output_js.getMatrix() ];
            print_matrices(matrices,1e-4, [ 'original','estimated' ]);
            console.log('___ Maximum difference = ',maxd[mode]);
        }

        console.log('Maxd=',maxd);
        let maxv=0.0;
        for (let i=0;i<maxd.length;i++) {
            if (maxv < maxd[i])
                maxv=maxd[i];
        }

        console.log('Max=',maxv);
        
        
        assert.equal(true,(maxv<0.001));
        
    });
 

});
